include/llbuild/Basic/RedirectedProcess.h

```diff
@@ -0,0 +1,68 @@
+//===- RedirectedProcess.h --------------------------------------*- C++ -*-===//
+//
+// This source file is part of the Swift.org open source project
+//
+// Copyright (c) 2014 - 2015 Apple Inc. and the Swift project authors
+// Licensed under Apache License v2.0 with Runtime Library Exception
+//
+// See http://swift.org/LICENSE.txt for license information
+// See http://swift.org/CONTRIBUTORS.txt for the list of Swift project authors
+//
+//===----------------------------------------------------------------------===//
+//
+// This file contains abstractions for creating a process and redirecting
+// input and output of that process.
+//
+//===----------------------------------------------------------------------===//
+
+#ifndef LLBUILD_BASIC_REDIRECTEDPROCESS_H
+#define LLBUILD_BASIC_REDIRECTEDPROCESS_H
+
+#include "llvm/ADT/SmallString.h"
+
+#include <mutex>
+#include <vector>
+
+namespace llbuild {
+namespace basic {
+namespace sys {
+class RedirectedProcess {
+  bool shouldCaptureOutput;
+  struct ProcessInfo *innerProcessInfo;
+
+public:
+  RedirectedProcess(bool shouldCaptureOutput);
+
+  bool openPipe();
+
+  bool execute(const char *path, bool setGroupFlags, char *const *args,
+               char *const *envp, std::mutex &spawnedProcessMutex);
```

Comment:
I'm worried this is a bit too literal of a port of the existing code and doesn't really abstract enough over the underlying implementation details (exposing the mutex, for example), and requiring raw C strings.

For the mutex, its possible the right abstraction would be one which can manage the entire set of spawned subprocesses -- that might also clean up some of the killing and cancellation logic in the execution queue...

I'm not sure about this, and I'm fine with starting with something less than ideal which is still an improvement and helps with Windows portability...